libvast/src/system/index.cpp

```diff
@@ -904,29 +905,20 @@ index(index_actor::stateful_pointer<index_state> self,
                               std::move(query), std::move(*worker));
       }
       VAST_VERBOSE("{} pushes query {} to the backlog", *self, query);
-      auto rp = self->make_response_promise<void>();
+      auto rp = self->make_response_promise<uuid, uint32_t, uint32_t>();
       self->state.backlog.emplace(std::move(query), rp);
       return rp;
     },
-    [self](atom::internal, vast::query query,
-           query_supervisor_actor worker) -> caf::result<void> {
+    [self](atom::internal, vast::query query, query_supervisor_actor worker)
+      -> caf::result<uuid, uint32_t, uint32_t> {
```

Comment:
Can you add a struct for this besides the `query` struct?

E.g.,

```cpp
struct query_response {
  uuid id;
  uint32_t num_partitions;
  uint32_t num_scheduled;
};
```

Reply:
I think a struct would help, but it feels more like a `cursor` or `handle` type object, because `response` suggests that it contains an answer to the query.

Reply:
I like cursor or progress.

Reply:
For "progress", a bit more information is missing; "status" is probably most neutral.

Reply:
Status is an already used term that I don't want to overload, so let's go with cursor.

Reply:
:slightly_frowning_face: I actually like the fact that this triple does not have a name. I'm reluctant to introduce such a struct here.

Reply:
> I actually like the fact that this triple does not have a name

Can you elaborate on why? When reading the code for the first time, how would you infer the semantics of this triple? The messaging API is the public interface and should be possible to understand *without* looking at the implementation.

Reply:
My issue is that this collection of values is very ad-hoc, it doesn't make sense to have it in any other context.

How would you feel about using a more complete comment at the interface declaration in `actors.hpp`? Something ala:
```c++
  // Initiates evaluation of a query.
  // @param query The query to evaluate.
  // @returns <evaluation_handle: uuid, candidate partitions: uint32_t, scheduled partitions: uint32_t>.
  //           evaluation_handle can be used to schedule additional partitions from the candidate set.
  //           candidate_partitions is the number of partitions that qualify for the query.
  //           scheduled_partitions is the number of partitions in the initial evaluation batch.
  caf::replies_to<query>::with<uuid, uint32_t, uint32_t>,
```

Reply:
I'd be happy with that. As long as a reader doesn't have to read code and can infer the semantics from proper typing (ideally) or documentation (acceptable), I think we're doing the right thing.

Reply:
Why not just introduce a struct like this, so we can _go-to definition_ at both the sending and the receiving site instead of manually needing to go to the definition in `vast/system/actors.hpp`, finding the correct handler, and then hoping that the documentation exists there? Documentation in that header is also much easier to get out of sync than documentation for a given type.

```cpp
/// Information returned by the index for ongoing queries to allow for pagination.
struct query_cursor {
  /// The handle of the query; needed to schedule additional partitions from the candidate set.
  uuid query_id = {};

  /// The number of partitions that qualify for the query.
  uint32_t num_candidate_partitions = {};

  /// The number of partitions scheduled for the query.
  uint32_t num_scheduled_partitions = {};

  auto inspect(auto& f, query_cursor& x) {
    return f(x.query_id, x.num_candidate_partitions, x.num_scheduled_partitions);
  }
};
```

I really didn't wanna make a fuss out of this, and mostly posted the comment because I thought we'd all agree that naming things is always good. Kinda surprised—I tend to always prefer naming things over using tuples.